domain/secret/state/state.go

```diff
@@ -112,6 +114,160 @@ WHERE  label = $secretOwner.label`
 	return nil
 }
 
+// CreateCharmApplicationSecret creates a secret onwed by the specified application,
+// returning an error satisfying [secreterrors.SecretAlreadyExists] if a secret
+// owned by the same application with the same label already exists.
+// It also returns an error satisfying [applicationerrors.ApplicationNotFound] if
+// the application does not exist.
+func (st State) CreateCharmApplicationSecret(ctx context.Context, version int, uri *coresecrets.URI, appName string, secret domainsecret.UpsertSecretParams) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	revisionUUID, err := uuid.NewUUID()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		dbSecretOwner := secretApplicationOwner{SecretID: uri.ID, Label: secret.Label}
+
+		selectApplicationUUID := `SELECT &M.uuid FROM application WHERE name=$M.name`
+		selectApplicationUUIDStmt, err := st.Prepare(selectApplicationUUID, sqlair.M{})
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		result := sqlair.M{}
+		err = tx.Query(ctx, selectApplicationUUIDStmt, sqlair.M{"name": appName}).Get(&result)
+		if err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				return applicationerrors.ApplicationNotFound
+			} else {
+				return errors.Annotatef(err, "looking up napplication UUID for %q", appName)
+			}
+		}
+		dbSecretOwner.ApplicationUUID = result["uuid"].(string)
+
+		if err := st.createSecret(ctx, tx, version, uri, secret, revisionUUID, st.checkApplicationSecretLabelExists(appName, dbSecretOwner.ApplicationUUID)); err != nil {
+			return errors.Annotatef(err, "inserting secret records for secret %q", uri)
+		}
+
+		if err := st.upsertSecretApplicationOwner(ctx, tx, dbSecretOwner); err != nil {
+			return errors.Annotatef(err, "inserting application secret record for secret %q", uri)
+		}
+		return nil
+	})
+	return errors.Trace(domain.CoerceError(err))
+}
+
+// checkApplicationSecretLabelExists returns function which checks if a charm application secret with the given label already exists.
+func (st State) checkApplicationSecretLabelExists(appName, app_uuid string) checkExistsFunc {
+	return func(ctx context.Context, tx *sqlair.TX, label string) error {
+		if label == "" {
+			return nil
+		}
+
+		checkLabelExistsSQL := `
+SELECT &secretApplicationOwner.secret_id
```

Comment:
The label has to be unique for the application-owned secrets and all the peer units-owned secrets.

Reply:
fixed